### DISCUSSION TITLE
Node reuse versus cloning

### DISCUSSION THREAD

**Discussion by ehoogeveen-medweb** [babel/babel#16314] [Q&A] [answered]
I'm writing a small plugin and I was wondering: Is it okay to reuse nodes?

For example, given input
```js
import { foo } from "path";
import { bar } from "path";
```
and output
```js
import { foo, bar } from "path";
```
Could I insert the specifier node for `bar` into the specifiers array for the first import statement, then remove the 2nd import statement? Or do I need to call `types.cloneDeepWithoutLoc` on the specifier node and insert that instead?

What about identifiers? Can they be reused without cloning?

I checked the handbook (and past discussions) but I don't see any mention of cloning or node reuse :)

**Comment by liuxingbaoyu** [comment]
Nodes can be moved but cannot be reused.
In addition, we use `t.cloneNode` by default, if you have no special needs.

**Comment by ehoogeveen-medweb** [comment]
Thanks! I didn't notice the 2nd parameter of `cloneNode`.

What does "moved" mean in this context? Is it okay as long as the original path is removed, or are there specific functions for moving nodes?

**Reply by liuxingbaoyu** [reply]
Yes! You're welcome!

**Reply by ehoogeveen-medweb** [reply]
Just to confirm, is that "Yes, it is okay as long as the original path is removed"? :)

**Reply by liuxingbaoyu** [reply]
Yes.

**Comment by ehoogeveen-medweb** [comment]
Thanks for the answers!

**Comment by ehoogeveen-medweb** [comment]
Oh, one more question (kind of unrelated). It seems that if I do the following, it also removes the associated binding:
```js
const scope = path1.scope; // path1.scope === path2.scope
console.log(Object.keys(scope.bindings));
path2.pushContainer('specifiers', specifier);
path1.get('specifiers')[index].remove();
console.log(Object.keys(scope.bindings));
```
shows the binding for the specifier being removed (regardless of whether I push first or remove first). Should I just splice the specifiers array instead?
```js
const scope = path1.scope; // path1.scope === path2.scope
console.log(Object.keys(scope.bindings));
path2.pushContainer('specifiers', specifier);
path1.node.specifiers.splice(index, 1);
console.log(Object.keys(scope.bindings));
```
shows the binding being preserved and I still see the import statement being adjusted, but I don't know if I'm messing up the state.

**Reply by liuxingbaoyu** [reply, marked as answer]
This is a long-standing problem with Babel, where current binding information is not always reliable.
In fact, the former should be recommended, but it may actually cause side effects. You can use it flexibly according to actual needs.
If you wish to use precise binding information, use `path.scope.crawl()` to refresh the binding information.

**Reply by nicolo-ribaudo** [reply]
An alternative (which is the best way perf-wise) is to register bingings when you add them to the ast. So in your code you would do
```js
const scope = path1.scope; // path1.scope === path2.scope
console.log(Object.keys(scope.bindings));
path2.pushContainer('specifiers', specifier);
path1.get('specifiers')[index].remove();
scope.registerDeclaration(path2);
console.log(Object.keys(scope.bindings));
```

Example: https://astexplorer.net/#/gist/32162ca61ba71d492333099d06c5abd9/latest

**Reply by ehoogeveen-medweb** [reply]
Oh, thanks! Indeed, that seems to work great :)
